### ISSUE TITLE
Devanagari fonts are not properly rendered after PDF conversion

### ISSUE THREAD

**Issue by rahool9** [dompdf/dompdf#969]
Devanagari fonts are not properly rendered after PDF conversion with by default fonts. If custom fonts are used it still has problem rendering. Please suggest solution urgently.

**Comment by bsweeney** [comment]
Have you loaded the font into dompdf? Or referenced it using `@font-face` CSS declaration? What does the result look like? Do you have a sample of the HTML?

**Comment by rahool9** [comment]
Yes I have loaded my custom font into dompdf but when converted PDF it showed '??? ' marks. So i decided to use default font . Default font showed basic language characters but text is not rendered properly.
Original Text
![image](https://cloud.githubusercontent.com/assets/2929858/8177129/147c49d4-141f-11e5-8081-b7eef1681023.png)
Converted PDF
![image](https://cloud.githubusercontent.com/assets/2929858/8177136/291a8f18-141f-11e5-9f29-e31a6ff2a9db.png)

**Comment by rahool9** [comment]
Yes I have loaded my custom font into dompdf but when converted PDF it showed '??? ' marks. So i decided to use default font . Default font showed basic language characters but text is not rendered properly.
Original Text
![image](https://cloud.githubusercontent.com/assets/2929858/8177129/147c49d4-141f-11e5-8081-b7eef1681023.png)
Converted PDF
![image](https://cloud.githubusercontent.com/assets/2929858/8177136/291a8f18-141f-11e5-9f29-e31a6ff2a9db.png)

**Comment by bsweeney** [comment]
? means that the encoding being used doesn't support the characters. It's likely that dompdf is using one of the core PDF fonts, which means that your custom font did not load correctly and a fallback is being used or you are not referencing the font correctly.

The weird spacing for your custom fonts would seem to indicate that the font metrics are incorrect or missing. Can you post some sample HTML? Including the relevant CSS.

The core PDF fonts (Helvetica, Times Roman, Courier) only support characters from the Windows ANSI encoding. We include the DejaVu fonts for wider support, but even these only support a subset of the Unicode character set (see the [PDF samples](http://dejavu-fonts.org/wiki/PDF_samples)).

Maybe start with something simpler. For example, the following seems to work:
```html
<!DOCTYPE html>
<html>
<head>
  <meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>
<style>
  @font-face {
    src: url("http://experimenting.in/css3/webfonts/hindifontsdemo/gargi.ttf") format('truetype');
    font-family: "gargi";
  }
  body {
    font-family: gargi, dejvu sans, sans-serif;
  }
</style>
</head>
<body>
  <p>यह सीएसएस वेब फोंट्स है| (Gargi)</p>
</body>
</html>
```

**Comment by rahool9** [comment]
I am using DOMPDF inside joomla. I have included the font in CSS. After converting into PDF my custom font is getting embedded in the PDF file
![image](https://cloud.githubusercontent.com/assets/2929858/8202397/289e9aa6-14f7-11e5-877a-957bb92c2795.png)

but the encoding is shown as ANSI.

I have enabled UNICODE in DOMPDF settings
![image](https://cloud.githubusercontent.com/assets/2929858/8202414/551cbe28-14f7-11e5-81ad-1b55d4f73b54.png)

& 

![image](https://cloud.githubusercontent.com/assets/2929858/8202419/5e50d65a-14f7-11e5-81f9-91ef7909aff8.png)

Also I have set the char type to UTF-8 while printing PDF.
![image](https://cloud.githubusercontent.com/assets/2929858/8202437/7ec1f086-14f7-11e5-9b6b-7a231134698b.png)

Don't know from where ANSI encoding is applied

**Comment by bsweeney** [comment]
That's a bit odd. dompdf appears to recognize the font but assumes it to be ANSI encoded. Preparing a font using the included tools in v0.6.x should always encode to Unicode. What does the relevant section of the dompdf_font_family_cache file look like? And what files do you have in your font folder?

**Comment by rahool9** [comment]
Tried everything but it is still taking encoding as ANSI.
mangal.afm file contents
![image](https://cloud.githubusercontent.com/assets/2929858/8299754/815372c8-199a-11e5-911c-3454562b9878.png)


font directory contents
![image](https://cloud.githubusercontent.com/assets/2929858/8299761/8e2a7712-199a-11e5-9021-f1ab8e1cd5c6.png)

dompdf_font_family_cache.dist.php file contents
![image](https://cloud.githubusercontent.com/assets/2929858/8299766/9d710cc2-199a-11e5-8efa-97a3274c4abe.png)

**Comment by bsweeney** [comment]
The .AFM files use Windows ANSI encoding. You need a .UFM file to get the Unicode metrics. Which version of dompdf are you using? What are you using to generate the font metrics file?

**Comment by rahool9** [comment]
Tried using UFM file converted usinf ttf2ufm conerter , but it doesn't generate PDF file when UFM is used. The server keeps busy till timeout & nothing happens. I am using 0.6.0 version.

**Comment by bsweeney** [comment]
You will need to use the UFM file since the AFM only supports Windows ANSI characters. Not sure why the document is timing out, could be something unrelated to the font but caused by the different spacing. Have you tried a simple "Hello World" document with the Devanagari font?

**Comment by rahool9** [comment]
Yes I have tried. It works well with AFM but times out with UFM. How am I suppose to resolve UFM issues?

**Comment by bsweeney** [comment]
Can you post the UFM created? Perhaps there's something wrong with it.

I [rendered some basic text](http://eclecticgeek.com/dompdf/debug.php?identifier=6d3e5b515dabcc8b2de37e937af6b363) using Mangal and another font and didn't have any issues so it's probably not a problem intrinsic to dompdf. Might just be a configuration issue.

**Comment by rahool9** [comment]
Here is the link for UFM file 
https://www.dropbox.com/s/olh12r2h8ajduaf/mangal.ufm?dl=0

**Comment by bsweeney** [comment]
That looks fine. Have you tried a simple document like the test I did?

**Comment by rahool9** [comment]
Yes but the server times out nothing happens.

**Comment by bsweeney** [comment]
I'm not sure what's going on, then. If PHP is timing out then it should report where this happens. You should be able to find the information in your PHP error log or by enabling PHP to send errors to the browser.

**Comment by rahool9** [comment]
I am getting two errors in error log when using UFM  
1. Undefined offset: 0 in \dompdf\include\style.cls.php on line 842
2. Division by zero in \dompdf\lib\class.pdf.php on line 3648

Second might be the one for indefinite loop. Code snippet for the same looks like this 
![image](https://cloud.githubusercontent.com/assets/2929858/8518742/a7cd3678-23ea-11e5-9c38-e5633c826120.png)

**Comment by bsweeney** [comment]
First, I couldn't find a version of dompdf that matches up with the line numbering there. At least, not a released version. So the version of dompdf you're using may be modified. You might try dropping in a fresh copy of dompdf 0.6.1 and see what happens.

Second, no font size was passed, which could be related to what's going on in the first notice about the undefined offset. You might try simplifying your stylesheet and see what happens, but it may also be addressed if you replace dompdf.

Third, if this were a timeout issue PHP should provide an error indicating the file/line number where it stopped. Not that this information will necessarily help, but it wouldn't hurt to know. Have you tried increasing the timeout? Are you using tables?

**Comment by rahool9** [comment]
Tried with 0.6.1 version as well. Working well with AFM not with UFM.
Can you post your HTML, CSS and DMPDF folder... Just to make sure that it has nothing to do with PHP settings.

**Comment by bsweeney** [comment]
You should be able to get enough info from my debug helper:
http://eclecticgeek.com/dompdf/debug.php?identifier=a00653164f3316b81533430f7255f5d5

**Comment by bsweeney** [comment]
Could this be related to issue #1002?

**Comment by satyajitgaikwad** [comment]
Can you please try following content in your debug hepler? May be the issue is with connected letters. They are not being rendered properly.
Please try - निर्णय क्षमता

**Comment by bsweeney** [comment]
@satyajitgaikwad I'm not sure if this is related. I think the problem here is with the metrics, but so far I haven't seen anything definitive helping identify the issue nor have I been able to reproduce it.

The incorrect ordering/connecting is *could* be related, but it seems likely to be due to something else. I think we may look into it in relation to issue #712.

**Comment by bsweeney** [comment]
ref #655
